// appium-support/build/lib/image-util.js
'use strict';

var _regeneratorRuntime = require('babel-runtime/regenerator')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _buffer = require('buffer');

var _pngjs = require('pngjs');

var _bluebird = require('bluebird');

var _bluebird2 = _interopRequireDefault(_bluebird);

/**
 * @typedef {Object} Region
 * @property {number} left - The offset from the left side
 * @property {number} top - The offset from the top
 * @property {number} width - The width
 * @property {number} height - The height
 */

var BYTES_IN_PIXEL_BLOCK = 4;
var SCANLINE_FILTER_METHOD = 4;

/**
 * Crop the image by given rectangle (use base64 string as input and output)
 *
 * @param {string} base64Image The string with base64 encoded image
 * @param {Region} rect The selected region of image
 * @return {string} base64 encoded string of cropped image
 */
function cropBase64Image(base64Image, rect) {
  var image;
  return _regeneratorRuntime.async(function cropBase64Image$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        context$1$0.next = 2;
        return _regeneratorRuntime.awrap(base64ToImage(base64Image));

      case 2:
        image = context$1$0.sent;

        cropImage(image, rect);
        context$1$0.next = 6;
        return _regeneratorRuntime.awrap(imageToBase64(image));

      case 6:
        return context$1$0.abrupt('return', context$1$0.sent);

      case 7:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
}

/**
 * Create a pngjs image from given base64 image
 *
 * @param {string} base64Image The string with base64 encoded image
 * @return {PNG} The image object
 */
function base64ToImage(base64Image) {
  var imageBuffer;
  return _regeneratorRuntime.async(function base64ToImage$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        imageBuffer = new _buffer.Buffer(base64Image, 'base64');
        return context$1$0.abrupt('return', new _bluebird2['default'](function (resolve, reject) {
          var image = new _pngjs.PNG({ filterType: SCANLINE_FILTER_METHOD });
          image.parse(imageBuffer, function (err, image) {
            // eslint-disable-line promise/prefer-await-to-callbacks
            if (err) {
              return reject(err);
            }
            resolve(image);
          });
        }));

      case 2:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
}

/**
 * Create a base64 string for given image object
 *
 * @param {PNG} image The image object
 * @return {string} The string with base64 encoded image
 */
function imageToBase64(image) {
  return _regeneratorRuntime.async(function imageToBase64$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        return context$1$0.abrupt('return', new _bluebird2['default'](function (resolve, reject) {
          var chunks = [];
          image.pack().on('data', function (chunk) {
            return chunks.push(chunk);
          }).on('end', function () {
            resolve(_buffer.Buffer.concat(chunks).toString('base64'));
          }).on('error', function (err) {
            // eslint-disable-line promise/prefer-await-to-callbacks
            reject(err);
          });
        }));

      case 1:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
}

/**
 * Crop the image by given rectangle
 *
 * @param {PNG} image The image to mutate by cropping
 * @param {Region} rect The selected region of image
 */
function cropImage(image, rect) {
  var imageRect = { width: image.width, height: image.height };
  var interRect = getRectIntersection(rect, imageRect);
  if (interRect.width < rect.width || interRect.height < rect.height) {
    throw new Error('Cannot crop ' + JSON.stringify(rect) + ' from ' + JSON.stringify(imageRect) + ' because the intersection between them was not the size of the rect');
  }

  var firstVerticalPixel = interRect.top;
  var lastVerticalPixel = interRect.top + interRect.height;

  var firstHorizontalPixel = interRect.left;
  var lastHorizontalPixel = interRect.left + interRect.width;

  var croppedArray = [];
  for (var y = firstVerticalPixel; y < lastVerticalPixel; y++) {
    for (var x = firstHorizontalPixel; x < lastHorizontalPixel; x++) {
      var firstByteIdxInPixelBlock = imageRect.width * y + x << 2;
      for (var byteIdx = 0; byteIdx < BYTES_IN_PIXEL_BLOCK; byteIdx++) {
        croppedArray.push(image.data[firstByteIdxInPixelBlock + byteIdx]);
      }
    }
  }

  image.data = new _buffer.Buffer(croppedArray);
  image.width = interRect.width;
  image.height = interRect.height;
  return image;
}

function getRectIntersection(rect, imageSize) {
  var left = rect.left >= imageSize.width ? imageSize.width : rect.left;
  var top = rect.top >= imageSize.height ? imageSize.height : rect.top;
  var width = imageSize.width >= left + rect.width ? rect.width : imageSize.width - left;
  var height = imageSize.height >= top + rect.height ? rect.height : imageSize.height - top;
  return { left: left, top: top, width: width, height: height };
}

exports.cropBase64Image = cropBase64Image;
exports.base64ToImage = base64ToImage;
exports.imageToBase64 = imageToBase64;
exports.cropImage = cropImage;
